### ISSUE TITLE
PKCS11/SoftHSM/RSA issue: Decryption with a different key to the one used for Encryption is allowed

### ISSUE THREAD

**Issue by tgonzalezorlandoarm** [parallaxsecond/parsec#761]
When trying to update the docker image [parsec-service-test-all](https://github.com/parallaxsecond/parsec/blob/06a98a3fc02ffa6ecb3bf90178294c8f758499dd/e2e_tests/docker_image/parsec-service-test-all.Dockerfile), the [asym_encrypt_decrypt_rsa_pkcs_different_keys](https://github.com/parallaxsecond/parsec/blob/06a98a3fc02ffa6ecb3bf90178294c8f758499dd/e2e_tests/tests/per_provider/normal_tests/asym_encryption.rs#L250) test fails for the PKCS11 provider.

This (negative) test encrypts a message with an RSA key and tries to decrypt it with another RSA key.
The RSA padding should allow the PKCS11 provider to realize that an incorrect key is being used and fail accordingly. The latter should happen when calling [decrypt for the PKCS11 session](https://github.com/parallaxsecond/parsec/blob/06a98a3fc02ffa6ecb3bf90178294c8f758499dd/src/providers/pkcs11/asym_encryption.rs#L72).

[SoftHSMv2](https://github.com/opendnssec/SoftHSMv2) is being used as the software backend for the PKCS11 provider. A [similar issue](https://github.com/opendnssec/SoftHSMv2/issues/739) has been reported in there, for RHEL9. We are using Ubuntu 22.04.

Unfortunately, [development in SoftHSMv2](https://github.com/opendnssec/SoftHSMv2/issues/712#issuecomment-2045162376) is currently dormant (since last month) so it does not seem likely that we'll get a solution from that end.

**Comment by tgonzalezorlandoarm** [comment]
@gowthamsk-arm @ionut-arm what are your ideas on this?

I guess the best option would be to find a new testing module for the PKCS11 provider, but we could ignore this particular test for the pkcs11 provider in the meantime?

**Comment by ionut-arm** [comment]
Sounds like a good plan! Maybe there's some alternative PKCS11 implementation that you can use if this doesn't get fixed in SoftHSM...

**Comment by tgonzalezorlandoarm** [comment]
Hi @anta5010 , thanks for the comment! We'll check with the new latest version and come back here

**Comment by tgonzalezorlandoarm** [comment]
@anta5010 latest version of SoftHSM fails as well

**Comment by tgonzalezorlandoarm** [comment]
@anta5010 Correct, if/when we change the implementation that we use, we'll need to update the OE QA tests in meta-parsec, thanks for the comment.

**Comment by tgonzalezorlandoarm** [comment]
Given the nature of the problem (a fixed SHA is failing when updating the docker image), we believe that this is due to the updates/changes in the toolchain, so it's not necessary linked to a change in OS.

**Comment by gowthamsk-arm** [comment]
@anta5010 I ran the automated PKCS11 tests on yocto and its passing. The failure on the CI is due to the libts library using new RPC calls

**Comment by gowthamsk-arm** [comment]
the current release of the parsec should fix the ci failure

**Comment by anta5010** [comment]
> @anta5010 I ran the automated PKCS11 tests on yocto and its passing. The failure on the CI is due to the libts library using new RPC calls

so, either Yocto toolchain or SoftHSM build parameters fix the pkcs11 issue you see in Parsec CI. If the latter then we can use the same approach in CI. Did you have a chance to have a look at SoftHSM Yocto recipe/version?

**Comment by gowthamsk-arm** [comment]
The yocto tests just run the cli tests as seen [here](https://gitlab.com/akuster/meta-security/-/blob/master-next/meta-parsec/lib/oeqa/runtime/cases/parsec.py#L199)
The failing test is part of e2e_tests in normal_tests, so its not running in the yocto.

**Comment by gowthamsk-arm** [comment]
Projects which include Parsec via meta-security, like Cassini, are currently seeing build failures due to the trusted service v1.0.0 update on meta-arm. Can we merge #764 for publishing the new parsec version to fix the TS issue and continue the discussion on this ticket in the meantime?

**Comment by anta5010** [comment]
I don't see any problem with that as long as we have this issue open :)

**Comment by anta5010** [comment]
> Projects which include Parsec via meta-security, like Cassini, are currently seeing build failures due to the trusted service v1.0.0 update on meta-arm.

Most likely they use scarthgap Yocto release. Please remember to update Parsec recipes in  scarthgap as well as in master. We used to time Parsec releases before Yocto release. But, not this time.

**Comment by gowthamsk-arm** [comment]
Yes I will send the patches to both master and scarthgap :)
